File: dotnet/Multiplayer/Rooms/Actions/SubmitAction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Michitai.Multiplayer.Rooms.Actions
{
    /// <summary>
    /// Request parameters for submitting an action to players in a room.
    /// Allows targeting specific players with typed action data.
    /// </summary>
    /// <typeparam name="T">The type of request data.</typeparam>
    public class SubmitAction<T> where T : class, new()
    {
        /// <summary>
        /// Which players to target with the action (default: All).
        /// </summary>
        public ERoomTargetPlayers Target_players { get; private set; } = ERoomTargetPlayers.All;

        /// <summary>
        /// Specific player IDs to target with the action.
        /// </summary>
        public int[]? Target_players_ids { get; private set; }

        /// <summary>
        /// The type of action being submitted.
        /// </summary>
        public string Action_type { get; private set; } = string.Empty;

        /// <summary>
        /// The request data for the action.
        /// </summary>
        public T? Request_data { get; private set; }

        /// <summary>
        /// Initializes a new SubmitAction request.
        /// </summary>
        /// <param name="targetPlayers">Which players to target with the action.</param>
        /// <param name="type">The type of action being submitted.</param>
        /// <param name="data">The request data for the action.</param>
        /// <param name="targetPlayersIds">Specific player IDs to target with the action.</param>
        public SubmitAction(ERoomTargetPlayers targetPlayers, string type, T? data = null, int[]? targetPlayersIds = null)
        {
            Target_players = targetPlayers;
            Target_players_ids = targetPlayersIds;
            Action_type = type;
            Request_data = data;
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/rooms/actions/SubmitAction.java
package com.michitai.multiplayer.rooms.actions;

/**
 * Request parameters for submitting an action to players in a room.
 * Allows targeting specific players with typed action data.
 *
 * @param <T> The type of request data.
 */
public class SubmitAction<T> {
    private ERoomTargetPlayers targetPlayers;
    private int[] targetPlayersIds;
    private String actionType;
    private T requestData;

    public SubmitAction(ERoomTargetPlayers targetPlayers, String actionType, T requestData, int[] targetPlayersIds) {
        this.targetPlayers = targetPlayers;
        this.targetPlayersIds = targetPlayersIds;
        this.actionType = actionType;
        this.requestData = requestData;
    }

    public SubmitAction(ERoomTargetPlayers targetPlayers, String actionType, T requestData) {
        this(targetPlayers, actionType, requestData, null);
    }

    public ERoomTargetPlayers getTargetPlayers() {
        return targetPlayers;
    }

    public void setTargetPlayers(ERoomTargetPlayers targetPlayers) {
        this.targetPlayers = targetPlayers;
    }

    public int[] getTargetPlayersIds() {
        return targetPlayersIds;
    }

    public void setTargetPlayersIds(int[] targetPlayersIds) {
        this.targetPlayersIds = targetPlayersIds;
    }

    public String getActionType() {
        return actionType;
    }

    public void setActionType(String actionType) {
        this.actionType = actionType;
    }

    public T getRequestData() {
        return requestData;
    }

    public void setRequestData(T requestData) {
        this.requestData = requestData;
    }
}
